File: Codes/1123-lowest-common-ancestor-of-deepest-leaves_20250405.java
/*
 * @Author: LetMeFly
 * @Date: 2025-04-05 00:59:21
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-04-05 01:03:59
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Pair<TreeNode, Integer> dfs(TreeNode root) {
        if (root == null) {
            return new Pair<>(root, 0);
        }
        Pair<TreeNode, Integer> left = dfs(root.left);
        Pair<TreeNode, Integer> right = dfs(root.right);
        if (left.getValue() > right.getValue()) {
            return new Pair<>(left.getKey(), left.getValue() + 1);
        } else if (left.getValue() < right.getValue()) {
            return new Pair<>(right.getKey(), right.getValue() + 1);
        }
        return new Pair<>(root, left.getValue() + 1);
    }

    public TreeNode lcaDeepestLeaves(TreeNode root) {
        return dfs(root).getKey();
    }
}

File: Codes/1123-lowest-common-ancestor-of-deepest-leaves_20250405.py
'''
Author: LetMeFly
Date: 2025-04-05 00:49:25
LastEditors: LetMeFly.xyz
LastEditTime: 2025-04-05 00:58:30
'''
from typing import Tuple, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def dfs(self, root: Optional[TreeNode]) -> Tuple[Optional[TreeNode], int]:
        if not root:
            return (root, 0)
        left = self.dfs(root.left)
        right = self.dfs(root.right)
        if left[1] > right[1]:
            return (left[0], left[1] + 1)
        elif left[1] < right[1]:
            return (right[0], right[1] + 1)
        return (root, left[1] + 1)
        
    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        return self.dfs(root)[0]